### ISSUE TITLE
Notification is Not working in Netopeer2 

### ISSUE THREAD

**Issue by vishalwc** [CESNET/Netopeer2#120]
When  send create-subscription from client as below, the rpc-reply "ok" is received from netopeer
<create-subscription xmlns="urn:ietf:params:xml:ns:netconf:notification:1.0"/>
rpc-reply ok from netopeer2

Further, we generated events from the device and the events are received at sysrepo but not received to client as it looks like Netopeer do not send it 
Initially the netopeer2-server sends capabilities to client  and notification capability  is present

can you please let us know Notification feature status ?

**Comment by michalvasko** [comment]
Hi,
notifications should be fully implemented, but were not yet properly merged. We have some problems with *sysrepo* merging, so please be patient.

Regards,
Michal

**Comment by vishalwc** [comment]
Thanks Michal for the quick update ,appreciate your support .

Just for you information as we might be picking the incomplete implementation 
- we tried this on notification branch 
- we also tried with op-notification.c in devel-server 

So if I have correctly understood notification are not merge properly with sysrepo and hence it is not sent out to client ?
please let us know if you plan to merge , we are in critical path for this .

**Comment by michalvasko** [comment]
Alright, but not everything is guaranteed to work. So, I would suggest you use *netopeer2-server* from `devel-server` branch with merged `notifications` branch and *sysrepo* from my [fork](https://github.com/michalvasko/sysrepo/tree/devel).

Regards,
Michal

**Comment by vishalwc** [comment]
small query to follow it up
The op_notification.c is modified in devel branch and hence with it on devel branch it is kind of merge with notification ,which is your first suggestion . was this correct merge ?

**Comment by vishalwc** [comment]
when are you planning to merge with sysrepo? we would prefer to wait for it .as we have to do installation of libraries on our devices .

**Comment by michalvasko** [comment]
Hi,
like I said, we have some problems with the merge, specifically with some tests. It is hard to say when it will be resolved, hopefully this week, but maybe not. `op_notification.c` does not exist in the `devel-server` branch of *netopeer2*, so I don't exactly know what you are talking about.

Regards,
Michal

**Comment by vishalwc** [comment]
Hi Michal
we Updated to sysrepo/devel to 2e5fd2a..578698f

Netopeer Init logs ( looks like netconf-server endpoints get request fails ) 
 
netopeer2-server[24170]: (cl_message_recv:211) Sysrepo server disconnected.
netopeer2-server[24170]: (cl_request_process:449) Unable to receive the message with response (session id=1326882673, operation=get-items).
**netopeer2-server[24170]: (sr_get_items_iter:1030) Sending get_items request failed '/ietf-netconf-server:netconf-server/listen/endpoint[ssh]//*' **
netopeer2-server[24170]: Failed to get "The peer disconnected" values iterator from sysrepo (`).
netopeer2-server[24170]: Server init failed.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=unsubscribe).
netopeer2-server[24170]: (cl_subscription_close:243) Error by processing of the request.
netopeer2-server[24170]: (cl_sm_cleanup:1756) Client Subscription Manager successfully destroyed.
netopeer2-server[24170]: (sr_unsubscribe:2845) Error by closing of the subscription: The peer disconnected.
netopeer2-server[24170]: (cl_message_send:170) Error by sending of the message: Broken pipe.
netopeer2-server[24170]: (cl_request_process:438) Unable to send the message with request (session id=1326882673, operation=session-stop).
netopeer2-server[24170]: (sr_session_stop:571) Error by processing of the request


sysrepo aborts during netopeer2 init   
**yang_read_module: Assertion `module' failed.
Aborted**

see the logs below 

[DBG] (cm_out_notif_process:1376) Sending a notification to '/tmp/sysrepo-subscriptions/ietf-netconf-server/24170.TgSG4c.sock'.
[DBG] (sm_connection_find_dst:573) Cannot find the connection with dst_address address='/tmp/sysrepo-subscriptions/ietf-netconf-server/24170.TgSG4c.sock'.
[DBG] (cm_out_notif_process:1391) Creating a new connection for the notification destination '/tmp/sysrepo-subscriptions/ietf-netconf-server/24170.TgSG4c.sock'
[DBG] (sm_connection_start:353) New connection started successfully, fd=16, conn ctx=0x6976d0.
[DBG] (cm_conn_buffer_expand:477) Output buffer for fd=16 expanded to 1024 bytes.
[DBG] (cm_conn_out_buff_flush:505) Sending 111 bytes of data.
[DBG] (cm_conn_out_buff_flush:511) 111 bytes of data sent.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=4, count=0.
[DBG] (rp_get_items_req_process:783) Processing get_items request.
[INF] (rp_dt_get_values_wrapper_with_opts:1355) Get items request startup datastore, xpath: /ietf-netconf-server:netconf-server/listen/endpoint[ssh]//*, offset: 0, limit: 100
[DBG] (ac_set_identity:155) Switching identity to UID='0' and GID='0' (username: root).
[DBG] (dm_load_data_tree_file:1122) Loaded module ietf-netconf-server: mtime sec=1491462550 nsec=836638489
[DBG] (dm_load_data_tree_file:1189) Usage count ietf-netconf-server incremented (value=1)
[INF] (dm_load_data_tree_file:1195) Data file /etc/sysrepo/data/ietf-netconf-server.startup loaded successfully
[DBG] (ac_set_identity:155) Switching identity to UID='0' and GID='0' (username: root).
[DBG] (dm_load_data_tree_file:1122) Loaded module ietf-keystore: mtime sec=1491462550 nsec=619638475
[DBG] (dm_module_clb:446) CALLBACK FOR MODULE (null) urn:ietf:params:xml:ns:yang:ietf-keystore
sysrepod: /mnt/local/satheeshk/git_open_prj/devel_branch/libyang/src/parser_yang.c:2680: 
yang_read_module: Assertion `module' failed.
Aborted

**Comment by michalvasko** [comment]
Hi,
yeah, I was fixing this yesterday, so please pull the most recent *libyang* and *netopeer2-server*, *sysrepo* too, while you at it.

Regards,
Michal

**Comment by vishalwc** [comment]
Thanks Michal will update and let you know how it goes .

**Comment by vishalwc** [comment]
Currently the callback function is not triggered, if notification is subscribed for whole module(xpath) with following sysrepo client_library call

sr_event_notif_subscribe_tree(np2srv.sr_sess.srs, path, np2srv_ntf_clb, NULL,
                                      SR_SUBSCR_NOTIF_REPLAY_FIRST | SR_SUBSCR_CTX_REUSE, &np2srv.sr_subscr);

Call back is triggered and notification is received on neconf clients, only when subscribed to individual notification (xpath) as below 

sr_event_notif_subscribe(np2srv.sr_sess.srs, "/abcd-efgh-interfaces:loam-loopback", np2srv_ntf_clb, NULL,
                                      SR_SUBSCR_DEFAULT, &subscription);

it is expected to work with module subscription as the controller or management entity will subscribe for it. would be helpful if you can provide the fix as earliest

**Comment by michalvasko** [comment]
Hi,
please try my referenced *sysrepo* fork or the pull request [sysrepo #787](https://github.com/sysrepo/sysrepo/pull/787) whether it solved your problem, thanks.

Regards,
Michal

**Comment by fabioantonini** [comment]
Hi Michal/Vishal
it works fine on my side
![screenshot from 2017-04-11 14-33-48](https://cloud.githubusercontent.com/assets/5917085/24909482/5e5a1386-1ec4-11e7-8169-19b0ccb37bc2.png)
regards
fabio

**Comment by fabioantonini** [comment]
it works also in netopeer2-cli

```
> subscribe 
OK
> 
notification (2016-05-30T12:55:22Z)
<netconf-session-end xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-notifications">
  <username>root</username>
  <session-id>3</session-id>
  <source-host>10.11.4.10</source-host>
  <termination-reason>closed</termination-reason>
</netconf-session-end>

> 
notification (2016-05-30T12:55:23Z)
<netconf-session-start xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-notifications">
  <username>root</username>
  <session-id>5</session-id>
  <source-host>10.11.4.10</source-host>
</netconf-session-start>
```

**Comment by vishalwc** [comment]
Hi Michal 

**Replay Issue** : 
After replay, all other existing/new subscriber clients stop receiving notification due to Netopeer2 didn’t receive callback.

Issued following command in netopeer-cli for replay
subscribe --stream NETCONF --begin 2017-04-13T14:27:00Z

The Sysrepo logs:

[DBG] (cm_conn_read_cb:1106) fd 18 readable
[DBG] (cm_conn_read_cb:1119) 992 bytes of data received on fd 18
[DBG] (cm_conn_read_cb:1129) fd 18 would block
[DBG] (cm_conn_in_buff_process:1059) New message of size 988 bytes received.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=7, current count=1.
[DBG] (rp_msg_process:4074) Threads: active=0/4, 1 requests in queue
[DBG] (rp_worker_thread_execute:3691) Thread id=140737299007232 signaled.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=8, count=0.
[DBG] (rp_event_notif_req_process:2973) Processing event notification request (/abcd-interfaces:loam-loopback).
[DBG] (dm_get_data_info_internal:2646) Module abcd-interfaces already loaded
[DBG] (dm_validate_procedure_content:5815) Validation succeeded for 'abcd-interfaces' module
[DBG] (pm_module_data_version_changed:577) Module 'abcd-interfaces' persist file version matches with cached value (1492091321842735672).
[DBG] (pm_get_subscriptions:1282) Returning 12 subscriptions from 'abcd-interfaces' persist file cache.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=1, current count=1.
[DBG] (cm_msg_enqueue_cb:1718) New message enqueued into CM message queue.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=2, count=0.
[DBG] (cm_out_event_notif_process:1547) Sending an event notification to '/tmp/sysrepo-subscriptions/abcd-interfaces/9854.UcU7eo.sock'.
[DBG] (cm_out_event_notif_process:1572) Reusing existing connection on fd=13 for the event notification destination '/tmp/sysrepo-subscriptions/abcd-interfaces/9854.UcU7eo.sock'
[DBG] (cm_conn_out_buff_flush:505) Sending 937 bytes of data.
[DBG] (cm_conn_out_buff_flush:511) 937 bytes of data sent.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=2, current count=1.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=3, current count=2.
[DBG] (rp_worker_thread_execute:3679) Thread id=140737299007232 will wait.
[DBG] (cm_msg_enqueue_cb:1718) New message enqueued into CM message queue.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=3, count=1.
[DBG] (cm_out_event_notif_process:1547) Sending an event notification to '/tmp/sysrepo-subscriptions/abcd-interfaces/9896.9h0ivB.sock'.
[DBG] (cm_out_event_notif_process:1572) Reusing existing connection on fd=27 for the event notification destination '/tmp/sysrepo-subscriptions/abcd-interfaces/9896.9h0ivB.sock'
[DBG] (cm_conn_out_buff_flush:505) Sending 1068 bytes of data.
[DBG] (cm_conn_out_buff_flush:511) 1068 bytes of data sent.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=4, count=0.
[DBG] (cm_conn_out_buff_flush:505) Sending 27 bytes of data.
[DBG] (cm_conn_out_buff_flush:511) 27 bytes of data sent.
[DBG] (cm_msg_enqueue_cb:1718) New message enqueued into CM message queue.

**Comment by vishalwc** [comment]
Hi Michal , 
when is the plan to merge your changes ( fork Michal ) with the devel ?

**Comment by michalvasko** [comment]
Hi,
regarding the replay issue, I was not able to reproduce it. I connected with CLI1, subscribed with simple `subscribe`. Then I connected with CLI2, CLI1 received a `netconf-session-start` notification. I then executed `subscribe --begin -30` (replay of last 30 seconds) in CLI2 and got mine `netconf-session-start` followed by `replayComplete`. Then I generated `link-discovered` notification from *sysrepo* test-module and received it in both CLI1 and CLI2. Can you please be more specific, what did you do and what did not work?

As for my changes, I have no *netopeer2* fork and if you mean my *sysrepo* fork, I push those changes regularly into *sysrepo* `devel` as pull-requests. Currently, I have no non-pushed changes. But if it happens, they should be merged in a day or two.

Regards,
Michal

**Comment by vishalwc** [comment]
Basic replay is working for us .keep you posted as we are moving ahead with testing notification more .

**Comment by vishalwc** [comment]
The sysrepo datastore for "ietf-netconf-notifications:netconf-config-change" is generated with "(!error!)" string  due to libyang context (ly_ctx) for the target configuration changed xpath is not available.
ly_ctx_get_module() always returns NULL when module name is "ietf-netconf-notifications"

The following is the generated xml file under  "/etc/sysrepo/data/notifications/ietf-netconf-notifications"

\<notifications xmlns="urn:ietf:params:xml:ns:yang:sysrepo-notification-store">
  \<notification>
    \<xpath>/ietf-netconf-notifications:netconf-config-change\</xpath>
    \<generated-time>2017-04-20T23:01:06-05:00\</generated-time>
    \<logged-time>3245838655\</logged-time>
    \<data>&lt;netconf-config-change xmlns=&quot;urn:ietf:params:xml:ns:yang:ietf-netconf-notifications&quot;&gt;&lt;changed-by&gt;&lt;session-id&gt;350575759&lt;/session-id&gt;&lt;username&gt;root&lt;/username&gt;&lt;/changed-by&gt;&lt;datastore&gt;running&lt;/datastore&gt;&lt;edit&gt;&lt;target xmlns:ks=&quot;urn:ietf:params:xml:ns:yang:ietf-keystore&quot;&gt;/ks:keystore&lt;/target&gt;&lt;operation&gt;delete&lt;/operation&gt;&lt;/edit&gt;&lt;edit&gt;&lt;target&quot;(!error!)&quot;&lt;operation&gt;merge&lt;/operation&gt;&lt;/edit&gt;&lt;/netconf-config-change&gt;\</data>
  \</notification>
\</notifications>

Due to above reason the replay fails with following error (sysrepo logs below)


[DBG] (cm_conn_read_cb:1106) fd 6 readable
[DBG] (cm_conn_read_cb:1119) 148 bytes of data received on fd 6
[DBG] (cm_conn_read_cb:1129) fd 6 would block
[DBG] (cm_conn_in_buff_process:1059) New message of size 144 bytes received.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=5, current count=1.
[DBG] (rp_msg_process:4074) Threads: active=0/4, 1 requests in queue
[DBG] (rp_worker_thread_execute:3691) Thread id=140737290614528 signaled.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=6, count=0.
[DBG] (rp_event_notif_replay_req_process:3132) Processing event notification replay request.
[DBG] (np_get_event_notifications:1747) Loading notifications '/ietf-netconf-notifications:*//.' generated between '30' and '1492747694'.
[DBG] (np_get_notification_files:548) Listing notification data files for 'ietf-netconf-notifications' modified from '1969-12-31 18:50' to '2017-04-20 23:18'.
[DBG] (np_get_notification_files:567) Adding file '/etc/sysrepo/data/notifications//ietf-netconf-notifications/2017-04-20_23-00.xml', mtim=1492747594
[DBG] (ac_set_identity:155) Switching identity to UID='0' and GID='0' (username: root).
[DBG] (sr_locking_set_lock_fd:789) File /etc/sysrepo/data/notifications//ietf-netconf-notifications/2017-04-20_23-00.xml has been locked
[DBG] (np_load_data_tree:410) Data successfully loaded from file '/etc/sysrepo/data/notifications//ietf-netconf-notifications/2017-04-20_23-00.xml'.
[DBG] (sr_locking_set_unlock_close_fd:858) File /etc/sysrepo/data/notifications//ietf-netconf-notifications/2017-04-20_23-00.xml (fd = 24) has been unlocked
[DBG] (dm_get_data_info_internal:2646) Module ietf-netconf-notifications already loaded
[DBG] (dm_ly_log_cb:1319) libyang error: Invalid NameStartChar of the attribute.
**[ERR] (dm_parse_event_notif:6105) Error by parsing of the notification XML tree: Invalid NameStartChar of the attribute.**
[DBG] (dm_ly_log_cb:1319) libyang error: Internal module "ietf-yang-metadata" cannot be removed.
[DBG] (dm_ly_log_cb:1319) libyang error: Internal module "yang" cannot be removed.
[DBG] (dm_ly_log_cb:1319) libyang error: Internal module "ietf-inet-types" cannot be removed.
[DBG] (dm_ly_log_cb:1319) libyang error: Internal module "ietf-yang-types" cannot be removed.
[DBG] (dm_ly_log_cb:1319) libyang error: Internal module "ietf-yang-library" cannot be removed.
[ERR] (np_get_event_notifications:1806) Error by parsing notification '/ietf-netconf-notifications:netconf-config-change'.
[ERR] (rp_event_notif_replay_req_process:3152) Error by loading event notifications for xpath '/ietf-netconf-notifications:*//.'.
[DBG] (sr_cbuff_enqueue:516) Circular buffer enqueue to position=5, current count=1.
[DBG] (rp_worker_thread_execute:3679) Thread id=140737290614528 will wait.
[DBG] (cm_msg_enqueue_cb:1718) New message enqueued into CM message queue.
[DBG] (sr_cbuff_dequeue:532) Circular buffer dequeue, new buffer head=6, count=0.
[DBG] (cm_conn_out_buff_flush:505) Sending 27 bytes of data.
[DBG] (cm_conn_out_buff_flush:511) 27 bytes of data sent

The following is the break point at failure while writing to 'ietf-netconf-notifications' notification store(upon generration of netconf-config-change notification):

Breakpoint 1, _transform_json2xml (module=0x7fffdc2ab260, expr=0x7fffec5943d0 "/ietf-system:system/abcd-efgh-system:system-config/abcd-efgh-system:clock/abcd-efgh-system:date", schema=0, 
    prefixes=0x7ffff4368940, namespaces=0x7ffff4368938, ns_count=0x7ffff4368934) at /devel_branch/libyang/src/common.c:373
                
The "abcd-efgh-system"              is not available  in ctx->models.list

**Comment by michalvasko** [comment]
Hi,
firstly, please update *sysrepo* because it's not the most recent version based on the output. But, it likely will not make a difference and not resolve your issue. Secondly, this notification is not always stored with an `(!error!)`, they worked fine for me, so please provide the schemas you use and the change you made. Also, I would be interested to see the backtrace from your breakpoint.

Regards,
Michal

**Comment by vishalwc** [comment]
This issue is always reproducible for us, on any config change.

The requested schemas for config change is given below:

notification (2017-04-25T05:18:59Z)
\<netconf-config-change xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-notifications">
  \<changed-by>
    \<session-id>1389391585\</session-id>
    \<username>root\</username>
  \</changed-by>
  \<datastore>running\</datastore>
  \<edit>
    \<target xmlns:if="urn:ietf:params:xml:ns:yang:ietf-interfaces" xmlns:efghif="http://www.xyz.com/ns/abcd-efgh-interfaces">/if:interfaces/if:interface[if:name='0.2']/efghif:interface-config/efghif:gigabit/efghif:output-rate\</target>
    \<operation>merge</operation>
  \</edit>
\</netconf-config-change>


The requested backtraces are given below

(gdb) b common.c:372 if strcmp(name,"abcd-efgh-system") == 0
Breakpoint 1 at 0x7ffff6cbe711: file ......./libyang/src/common.c, line 372.
(gdb) c
Continuing.

Breakpoint 1, _transform_json2xml (module=0x7fffdc83da80, expr=0x7fffe401fea0 "/ietf-system:system/abcd-efgh-system:system-config/abcd-efgh-system:clock/abcd-efgh-system:date", schema=0, 
    prefixes=0x7ffff4b69940, namespaces=0x7ffff4b69938, ns_count=0x7ffff4b69934) at ......./libyang/src/common.c:372
372         ......./libyang/src/common.c
(gdb) bt
#0  _transform_json2xml (module=0x7fffdc83da80, expr=0x7fffe401fea0 "/ietf-system:system/abcd-efgh-system:system-config/abcd-efgh-system:clock/abcd-efgh-system:date", schema=0, 
    prefixes=0x7ffff4b69940, namespaces=0x7ffff4b69938, ns_count=0x7ffff4b69934) at ......./libyang/src/common.c:372
#1  0x00007ffff6cbf04a in transform_json2xml (module=0x7fffdc83da80, expr=0x7fffe401fea0 "/ietf-system:system/abcd-efgh-system:system-config/abcd-efgh-system:clock/abcd-efgh-system:date", 
    prefixes=0x7ffff4b69940, namespaces=0x7ffff4b69938, ns_count=0x7ffff4b69934) at ......./libyang/src/common.c:519
#2  0x00007ffff6d79adb in xml_print_leaf (out=0x7ffff4b69b50, level=0, node=0x7fffe401f3e0, toplevel=0, options=0) at ......./libyang/src/printer_xml.c:331
#3  0x00007ffff6d7a5dd in xml_print_node (out=0x7ffff4b69b50, level=0, node=0x7fffe401f3e0, toplevel=0, options=0) at ......./libyang/src/printer_xml.c:542
#4  0x00007ffff6d7a0ea in xml_print_list (out=0x7ffff4b69b50, level=0, node=0x7fffe401f3a0, is_list=1, toplevel=0, options=0) at ......./libyang/src/printer_xml.c:449
#5  0x00007ffff6d7a62c in xml_print_node (out=0x7ffff4b69b50, level=0, node=0x7fffe401f3a0, toplevel=0, options=0) at ......./libyang/src/printer_xml.c:548
#6  0x00007ffff6d79ea5 in xml_print_container (out=0x7ffff4b69b50, level=0, node=0x7fffe4020b10, toplevel=1, options=0) at ......./libyang/src/printer_xml.c:415
#7  0x00007ffff6d7a5bc in xml_print_node (out=0x7ffff4b69b50, level=0, node=0x7fffe4020b10, toplevel=1, options=0) at ......./libyang/src/printer_xml.c:539
#8  0x00007ffff6d7a8d9 in xml_print_data (out=0x7ffff4b69b50, root=0x7fffe4020b10, options=0) at ......./libyang/src/printer_xml.c:607
#9  0x00007ffff6d4fcb9 in lyd_print_ (out=0x7ffff4b69b50, root=0x7fffe4020b10, format=LYD_XML, options=0) at ......./libyang/src/printer.c:351
#10 0x00007ffff6d4fe07 in lyd_print_mem (strp=0x7ffff4b69c48, root=0x7fffe4020b10, format=LYD_XML, options=0) at ......./libyang/src/printer.c:408
#11 0x00007ffff7b01ee9 in dm_netconf_config_change_to_string (dm_ctx=0x633740, notif=0x7fffe401f2c0, out=0x7ffff4b69c48) at ......./sysrepo/src/data_manager.c:6661
#12 0x00007ffff7b0f679 in np_store_event_notification (np_ctx=0x604b10, user_cred=0x0, xpath=0x7fffec0f6be0 "/ietf-netconf-notifications:netconf-config-change", generated_time=1493086034, 
    notif_data_tree=0x7fffe401f2c0) at ......./sysrepo/src/notification_processor.c:1697
#13 0x00007ffff7a98bf9 in rp_event_notif_req_process (rp_ctx=0x604870, session=0x0, msg=0x7fffec0f7720) at ......./sysrepo/src/request_processor.c:3020
#14 0x00007ffff7a9b618 in rp_req_dispatch (rp_ctx=0x604870, session=0x0, msg=0x7fffec0f7720, skip_msg_cleanup=0x7ffff4b6bebb)
    at ......./sysrepo/src/request_processor.c:3391
#15 0x00007ffff7a9c856 in rp_msg_dispatch (rp_ctx=0x604870, session=0x0, msg=0x7fffec0f7720) at ......./sysrepo/src/request_processor.c:3539
#16 0x00007ffff7a9d1b8 in rp_worker_thread_execute (rp_ctx_p=0x604870) at ......./sysrepo/src/request_processor.c:3640
#17 0x00007ffff75dbdc5 in start_thread () from /lib64/libpthread.so.0
#18 0x00007ffff69e4ced in clone () from /lib64/libc.so.6
(gdb)

**Comment by michalvasko** [comment]
Hi,
the referenced pull request should fix this.

Regards,
Michal

**Comment by anjansat** [comment]
Hi Michal,

The replay works fine, after updating your fix for replay (e2055a6). But afterwards, all other existing/new subscriber clients stop receiving notification due to Netopeer2 didn’t receive callback.
I could see this issue continuously and as of now restarting netopeer2, resolves issue(re-subscription).

Regards
Anjan

**Comment by michalvasko** [comment]
Hi Anjan,
I am not sure if I understand it correctly, is there still an issue or not? If so, then please describe it better and you should even start a new issue, it has nothing to do with this one.

Regards,
Michal
